proposals/shared-everything-threads/Overview.md

```diff
@@ -616,24 +616,23 @@ now interpret as a u32:ordering immediate, allowing us to express release-acquir
 The shared annotation occupies the same bit location as it does for memory types. In the threads
 proposal, the `limits` flag byte on memories is extended such that if bit 1 (the second bit) is set,
 the memory is `shared`. Likewise:
-- tables are `shared` if the `limit` flag byte has bit 1 set
-- globals are `shared` if the `mut` flag byte has bit 1 set
-
-| type | opcode |
-| ---- | ------ |
-| `func shared [valtype*] -> [valtype*]` | 0x5d (-35 as s7) |
-| `struct shared fieldtype*` | 0x5c (-36 as s7) |
-| `array shared fieldtype` | 0x5b (-37 as s7) |
-| `(shared nofunc)` | 0x5A |
-| `(shared noextern)` | 0x59 |
-| `(shared none)` | 0x58 |
-| `(shared func)` | 0x57 |
-| `(shared extern)` | 0x56 |
-| `(shared any)` | 0x55 |
-| `(shared eq)` | 0x54 |
-| `(shared i31)` | 0x53 |
-| `(shared struct)` | 0x52 |
-| `(shared array)` | 0x51 |
+
+ - tables are `shared` if the `limit` flag byte has bit 1 set
+ - globals are `shared` if the `mut` flag byte has bit 1 set
+ - element segments are `shared` if bit 3 of their first u32 field is set (bits
+   0-2 are already used).
+ - data segments are `shared` if bit 2 of their first u32 fields is set (bits
+   0-1 are already used).
+ - tags are shared if their first byte is 0x01 (it is currently reserved and
+   required to be 0x0).
+
+| type                                   | opcode             |
+|----------------------------------------|--------------------|
+| `func shared [valtype*] -> [valtype*]` | 0x5d (-35 as s7)   |
+| `struct shared fieldtype*`             | 0x5c (-36 as s7)   |
+| `array shared fieldtype`               | 0x5b (-37 as s7)   |
```

Comment:
I would suggest that we also use a prefix byte here, both for consistency and because we may likewise add more composite types over time. We can reuse the same opcode as for abstypes.

(You also want to change the left column to have `shared` as a prefix, to much the rest of the document.)

Reply:
Using a prefix byte here makes sense to me.

For the text format, I would prefer to have e.g. `(func shared ...` over `(shared func ...` so that what is being parsed is clear from the beginning of the form. Otherwise we would have to read ahead past the `(shared ...` to figure out what we are parsing. OTOH I'm fine with e.g. `(shared anyref)` for the abstract heap types since their unshared versions don't have parens at all and there is only a single token after `shared`.

Reply:
The syntax is `shared comptype`, following previous discussion, the document itself [says so](https://github.com/WebAssembly/shared-everything-threads/blob/3991407851151555e5897b1f6baed6a43eb43903/proposals/shared-everything-threads/Overview.md?plain=1#L191) (but then writes it inconsistently, even within the same section ;) ). Hence there shouldn't be any parsing issue?

Reply:
Right, I think that makes sense for the abstract syntax, but it seems unfortunate for the text syntax. Maybe it's not worth having them diverge, though.

Reply:
Well, the text format is meant to represent the abstract syntax (plus a bit of sugar).